snorkel/contrib/fonduer/fonduer/candidates.py

```diff
@@ -1,21 +1,173 @@
+        super(CandidateExtractor, self).__init__(CandidateExtractorUDF,
+                                                 candidate_class=candidate_class,
+                                                 cspaces=cspaces,
+                                                 matchers=matchers,
+                                                 candidate_filter=candidate_filter,
+                                                 self_relations=self_relations,
+                                                 nested_relations=nested_relations,
+                                                 symmetric_relations=symmetric_relations)
+
+    def apply(self, xs, split=0, **kwargs):
+        """Call the CandidateExtractorUDF."""
+        super(CandidateExtractor, self).apply(xs, split=split, **kwargs)
+
+    def clear(self, session, split, **kwargs):
+        """Delete Candidates from given split the database."""
+        session.query(Candidate).filter(Candidate.split == split).delete()
+
+
+class CandidateExtractorUDF(UDF):
+    """UDF for performing candidate extraction."""
+
+    def __init__(self, candidate_class, cspaces, matchers, candidate_filter, self_relations, nested_relations, symmetric_relations, **kwargs):
+        """Initialize the CandidateExtractorUDF."""
+        self.candidate_class     = candidate_class
+        self.candidate_spaces    = cspaces if type(cspaces) in [list, tuple] else [cspaces]
+        self.matchers            = matchers if type(matchers) in [list, tuple] else [matchers]
+        self.candidate_filter    = candidate_filter
+        self.nested_relations    = nested_relations
+        self.self_relations      = self_relations
+        self.symmetric_relations = symmetric_relations
+
+        # Check that arity is same
+        if len(self.candidate_spaces) != len(self.matchers):
+            raise ValueError("Mismatched arity of candidate space and matcher.")
+        else:
+            self.arity = len(self.candidate_spaces)
+
+        # Make sure the candidate spaces are different so generators aren't expended!
+        self.candidate_spaces = map(deepcopy, self.candidate_spaces)
+
+        # Preallocates internal data structures
+        self.child_context_sets = [None] * self.arity
+        for i in range(self.arity):
+            self.child_context_sets[i] = set()
+
+        super(CandidateExtractorUDF, self).__init__(**kwargs)
+
+    def apply(self, context, clear, split, **kwargs):
+        """Extract candidates from the given Context.
+
+        Here, we define a context as a Phrase.
+        :param context:
+        :param clear:
+        :param split: Which split to use.
+        """
+        # Generate TemporaryContexts that are children of the context using the candidate_space and filtered
+        # by the Matcher
+        for i in range(self.arity):
+            self.child_context_sets[i].clear()
+            for tc in self.matchers[i].apply(self.candidate_spaces[i].apply(self.session, context)):
```

Comment:
Besides moving this code to being under Fonduer contrib, this is the only real change.

```py
for tc in self.matchers[i].apply(self.candidate_spaces[i].apply(self.session, context)):
```
Now passes in `self.session`, where it didn't before.